### PR TITLE
[SYCL] Support SYCL layer for LLaMA2 model

### PR DESCRIPTION
**Pull request by ThanatosShinji**
## Type of Change

- Convert BesTLA int4 weight to SYCL weight
- Load SYCL weight to ne_tensor
- Activation host->device conversion
- Device workspace design
- Convert 'output.weight' to SYCL tensor, run LLaMa2  with ARC750

### CODE REVIEW

**Comment by luoyu-intel** [comment]
SYCL UT passed and llama2 sycl model_eval with i7-1185G7's Iris iGPU, 12900K(as CPU device) and A770

**Comment by luoyu-intel** [comment]
QKV+FFN on GPU, others on CPU, 9600K+A770:
from 150ms/token to 110ms/token. memcpy between the device and host took the most time.

File: neural_speed/core/CMakeLists.txt [hunk 1]
```diff
@@ -36,9 +36,10 @@ endif()
 if(NOT WIN32)
   target_link_libraries(ne_layers PUBLIC rt)
 else()
-  target_link_options(ne_layers PUBLIC /STACK:5242880)
+  target_link_options(ne_layers PUBLIC /STACK:5242880 /F5242880)
```
Comment (zhentaoyu) [inline neural_speed/core/CMakeLists.txt:39]: Sorry, what does `/STACK:5242880 /F5242880` mean?

File: neural_speed/core/layers/ne_bestla.cpp [hunk 2]
```diff
@@ -162,3 +162,128 @@ void bestla_add(int batch, int vsize, const float* tensor, const float* vector,
...
+
+static inline int ne_nrows(const struct ne_tensor* tensor) {
+  static_assert(NE_MAX_DIMS == 4, "NE_MAX_DIMS is not 4 - update this function");
+  return tensor->ne[1] * tensor->ne[2] * tensor->ne[3];
+}
+
+ne_backend bestla_backend_support(struct ne_tensor* src0, struct ne_tensor* src1, enum ne_op op) {
+  ne_backend bk = NE_BACKEND_CPU;
+#ifdef NS_SYCL
+  bool src_on_devce = src0->backend == NE_BACKEND_SYCL;
```
Comment (zhentaoyu) [inline neural_speed/core/layers/ne_bestla.cpp:179]: typo? `src_on_device`?

File: neural_speed/core/CMakeLists.txt (hunk 1 above)
Comment (luoyu-intel) [reply neural_speed/core/CMakeLists.txt:39]: enlarge the stack size, /STACK for MSVC, /F for DPCPP

File: neural_speed/core/layers/ne_bestla.cpp (hunk 2 above)
Comment (luoyu-intel) [reply neural_speed/core/layers/ne_bestla.cpp:179]: nice catch

Comment (luoyu-intel) [reply neural_speed/core/layers/ne_bestla.cpp]: fixed

**Comment by ThanatosShinji** [comment]
Disable CPU MHA, run all other layers on GPU A750-8GB, 15 ms/token

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by ThanatosShinji** [comment]
Run LLaMa2 all layers on A750: 19.5ms/token
MHA+Rope+RMS=5ms, need further optimization.

**Comment by ThanatosShinji** [comment]
n_context = 1024, in=512, out=512, only 5.8GB GPU memory.

**Comment by ThanatosShinji** [comment]
what is your name? what is your favorite color?
what is your name? what is your favorite
model_print_timings:        load time =  2672.97 ms
model_print_timings:      sample time =     2.92 ms /    16 runs   (    0.18 ms per token)
model_print_timings: prompt eval time =  2672.43 ms /     6 tokens (  445.40 ms per token)
model_print_timings:        eval time =   258.97 ms /    15 runs   (   17.26 ms per token)
model_print_timings:       total time =  2936.64 ms
========== eval time log of each prediction ==========
prediction   0, time: 2672.43ms
prediction   1, time: 17.88ms
prediction   2, time: 17.26ms
prediction   3, time: 17.33ms
prediction   4, time: 17.85ms

**Comment by ThanatosShinji** [comment]
MTL, 54ms/ token, the total latency of int4 gemm is 47ms, optimize rope and layer norm, implement f16 MHA (flash attention) should help reduce this latency.

**Review by luoyu-intel — APPROVED** [review]

**Comment by luoyu-intel** [comment]
llama2-7b on A770+9900K:

```
 Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have exciting experiences. However, no matter how hard she tried, she always got bored and restless wherever she went.
One day, while wandering
model_print_timings:        load time =   231.00 ms
model_print_timings:      sample time =     6.60 ms /    32 runs   (    0.21 ms per token)
model_print_timings: prompt eval time =   230.26 ms /    32 tokens (    7.20 ms per token)
model_print_timings:        eval time =   523.42 ms /    31 runs   (   16.88 ms per token)
model_print_timings:       total time =   765.82 ms
========== eval time log of each prediction ==========
prediction   0, time: 230.26ms
prediction   1, time: 17.19ms
prediction   2, time: 16.78ms
prediction   3, time: 16.82ms
prediction   4, time: 16.65ms
```

**Pull request merged**